rust/src/delta.rs

```diff
@@ -540,8 +550,49 @@ impl DeltaTable {
             self.version(),
         );
 
-        while let PeekCommit::New(new_version, actions) =
-            self.peek_next_commit(self.version()).await?
+        let buf_size = self.config.log_buffer_size;
+        if buf_size == 0 {
+            return Err(DeltaTableError::Generic(String::from("Log buffer size cannot be zero!")))
+        }
+
+        // construct stream yielding (version, bytes)
+        let store = self.storage.clone();
+        let log_stream = futures::stream::iter(self.version() + 1..).map(|i| {
+            let store2 = store.clone();
+            let loc = commit_uri_from_version(i).clone();
+            async move {
+                let out = store2.get(&loc);
+                (i, out.await)
+            }
+        });
+        
+        let mut log_buffer = {
+            match max_version {
+                None => {
+                    log_stream.take(usize::MAX).buffered(buf_size)
+                }
+                Some(n) => {
+                    if n - self.version() < 0 {
+                        return Err(DeltaTableError::Generic(String::from("Table version is greater than max_version!")));
+                    }
+                    let n_commits = usize::try_from(n - self.version() + 2);
+                    match n_commits {
+                        Ok(n) => log_stream.take(n).buffered(buf_size),
+                        Err(err) => return Err(DeltaTableError::GenericError { source: Box::new(err) })
+                    }
+                }
+            }
+        };
+        
+        while let Some((new_version, actions)) = {
+            let next_commit = log_buffer.next().await;
+            match next_commit {
+                Some((v, Ok(x))) => Ok(Some((v, self.get_actions(v,  x.bytes().await?).await?))),
+                Some((_, Err(ObjectStoreError::NotFound { .. }))) => Ok(None),
+                Some((_, Err(err))) => Err(DeltaTableError::GenericError { source: Box::new(err) }),  // TODO ??
+                None => Err(DeltaTableError::Generic(String::from("Log stream closed unexpectedly!")))
```

Comment:
My understanding is perhaps a bit hazy, but isn't a `None` at the end of the stream an expected condition? Won't this always happen once the stream is at the end?

Reply:
Yeah you're right, this is expected. I made this an Ok case and also updated the `.take` call to take `max_version - self.version()` elements instead of +2. Had some confusion there earlier due to this error.